hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/common/utils/TestSlidingWindow.java

```diff
@@ -0,0 +1,157 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.ozone.container.common.utils;
+
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import java.time.Duration;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.Timeout;
+
+/**
+ * Tests for {@link SlidingWindow} class.
+ */
+public class TestSlidingWindow {
+
+  private SlidingWindow slidingWindow;
+
+  @BeforeEach
+  public void setup() {
+    slidingWindow = new SlidingWindow(3, Duration.ofSeconds(5));
+  }
+
+  @Test
+  public void testConstructorValidation() {
+    // Test invalid window size
+    assertThrows(IllegalArgumentException.class, () -> 
+        new SlidingWindow(0, Duration.ofMillis(100)));
+    assertThrows(IllegalArgumentException.class, () -> 
+        new SlidingWindow(-1, Duration.ofMillis(100)));
+
+    // Test invalid expiry duration
+    assertThrows(IllegalArgumentException.class, () -> 
+        new SlidingWindow(1, Duration.ofMillis(0)));
+    assertThrows(IllegalArgumentException.class, () -> 
+        new SlidingWindow(1, Duration.ofMillis(-1)));
+  }
+
+  @Test
+  public void testAdd() {
+    for (int i = 0; i < slidingWindow.getWindowSize(); i++) {
+      slidingWindow.add();
+      assertFalse(slidingWindow.isFull());
+    }
+
+    slidingWindow.add();
+    assertTrue(slidingWindow.isFull());
+  }
+
+  @Test
+  public void testEventExpiration() throws InterruptedException {
+    slidingWindow = new SlidingWindow(2, Duration.ofMillis(500));
+
+    // Add events to reach threshold
+    slidingWindow.add();
+    slidingWindow.add();
+    slidingWindow.add();
+    assertTrue(slidingWindow.isFull());
+
+    // Wait for events to expire
+    Thread.sleep(600);
+
+    assertFalse(slidingWindow.isFull());
+
+    // Add one more event - should not be enough to mark as full
+    slidingWindow.add();
+    assertFalse(slidingWindow.isFull());
+  }
+
+  @Test
+  public void testPartialExpiration() throws InterruptedException {
+    slidingWindow = new SlidingWindow(3, Duration.ofSeconds(1));
+
+    slidingWindow.add();
+    slidingWindow.add();
+    slidingWindow.add();
+    slidingWindow.add();
+    assertTrue(slidingWindow.isFull());
+
+    Thread.sleep(600);
+    slidingWindow.add(); // this will remove the oldest event as the window is full
+
+    // Wait for the oldest events to expire
+    Thread.sleep(500);
+    assertFalse(slidingWindow.isFull());
+  }
+
+  @Test
+  @Timeout(value = 10)
+  public void testConcurrentAccess() throws InterruptedException {
```

Comment:
I don't think we need this test case.  `SlidingWindow` is a simple class, and we can also rely on static analysis (Spotbugs).